Adding debounce time to avoid spamming changelog creation

zocoi opened the pull request:
Usecase: We have a `Blog` model which get frequent minor updates for the `content` column as the editor keeps typing with the CMS autosave feature (every 5s) and we don't want to create log to capture tiny changes. The content text could be big so we need to save space. I came up with a concept similar to https://underscorejs.org/#debounce. The idea is that if the several logs came in within a time period, we only keep the latest one and discard the rest. 

```
{
    "h": [
        {
            "c": {
                "content": "Content 1"
            },
            "v": 1,
            "ts": 0
        },
        {
            "c": {
                "content": "content 2"
            },
            "v": 2,
            "ts": 100
        },
        {
            "c": {
                "content": "content 3"
            },
            "v": 3,
            "ts": 101
        }
    ],
    "v": 3
}
```

becomes

```
{
    "h": [
        {
            "c": {
                "content": "Content 1"
            },
            "v": 1,
            "ts": 0
        },
        {
            "c": {
                "content": "content 3"
            },
            "v": 3,
            "ts": 101
        }
    ],
    "v": 3
}
```

Notice that version 2 is gone

Please give feedback and help optimize the sql part.

Right now the code doesn't care if columns are edited by different `responsible_id`s but that could be easily added.

palkan commented on lib/generators/logidze/install/templates/migration.rb.erb:
```diff
@@ -205,6 +244,11 @@ class <%= @migration_class_name %> < ActiveRecord::Migration<%= ActiveRecord::VE
             END IF;
+
+            IF debounce_time IS NOT NULL THEN
+              NEW.log_data := logidze_merge_recent(NEW.log_data, debounce_time);
```
I think, it's better to handle debounce before adding a new version instead of adding a new first version first and than merging it with the previous.

So, on line 231 we can do:

```
IF debounce_time IS NOT NULL AND version->>'ts'::bigint - NEW.log_data#>'{h,-1,ts}')::text::bigin <= debounce_time
  # merge new version with the previous one
  version = logidze_merge_version(NEW.log_data#>'{h,-1}', version)
  # remove the previous version from log
  NEW.log_data = jsonb_set(
            NEW.log_data,
            '{h}',
             NEW.log_data->'h' - (size -1 )::text
          );
END
```

zocoi replied:
thanks for the brilliant suggestion, I made the change accordingly and added some tests

zocoi commented on lib/generators/logidze/install/templates/migration.rb.erb:
```diff
@@ -189,6 +192,25 @@ class <%= @migration_class_name %> < ActiveRecord::Migration<%= ActiveRecord::VE
+              (version->>'ts')::bigint - (NEW.log_data#>'{h,-1,ts}')::text::bigint <= debounce_time
+            ) THEN
+              -- merge new version with the previous one
+              version := NEW.log_data#>'{h,-1}' || version;
```
this will merge the `v`, `ts`, (meta) `m` and (responsible_id) `r` along with the changesets `c`. However, `||` can't do the deep merge for the changeset so we need to do it separate in the next line.

For meta `m`, we can do a deep merge but just like you said, I don't think it's neccesary

palkan commented:
@zocoi Let's add a changelog entry and update the readme with the new feature description. And we're almost done)

zocoi commented:
@palkan @charlie-wasp I changed the version number logics when merging logs, I kept the previous version number for consistency. Previous `1->2->3` was merged to `1->3`, now it should be `1->2`.
Also added changelog and updated README, let me know what else should be done.

Thanks

palkan commented on CHANGELOG.md:
```diff
@@ -2,6 +2,33 @@
 
 ## master
 
+## 0.8.0 (2018-10-01)
```
Let's leave the change log entry under `master` section. We don't know the date and version of the release  this feature will be included)

palkan commented on CHANGELOG.md:
```diff
@@ -2,6 +2,33 @@
 
+## 0.8.0 (2018-10-01)
+
+- [PR [#87](https://github.com/palkan/logidze/pull/87)] Adding debounce time to avoid spamming changelog creation
```
Please, add your Github handle here (and in the bottom of the document)

zocoi replied:
thanks, it should be done

palkan commented on CHANGELOG.md:
````diff
@@ -2,6 +2,90 @@
+
+without `debounce_time`
+
+```
````
Add `js` lang attributes here and in the snippet below to make RuboCop ignore these snippets (right it considers them as Ruby code)

palkan commented:
Great job! Thanks!

Will release a new version soon
